src/gt4sd/algorithms/generation/diffusion/implementation.py

```diff
@@ -0,0 +1,162 @@
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+
+def set_seed(seed: int = 42) -> None:
+    """Set seed for all random number generators.
+
+    Args:
+        seed: seed to set. Defaults to 42.
+    """
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+    if torch.cuda.is_available:
+        torch.cuda.manual_seed_all(seed)  # type:ignore
+
+
+MODEL_TYPES = {
+    "diffusion": DDPMPipeline,
+    "diffusion_implicit": DDIMPipeline,
+    "latent_diffusion": LDMPipeline,
+    "latent_diffusion_conditional": LDMTextToImagePipeline,
+    "stable_diffusion": StableDiffusionPipeline,
+    "score_sde": ScoreSdeVePipeline,
+}
+
+SCHEDULER_TYPES = {
+    "ddpm": DDPMScheduler,
+    "ddim": DDIMScheduler,
+    "discrete": LMSDiscreteScheduler,
+    "continuous": ScoreSdeVeScheduler,
+}
+
+
+class Generator:
+    """Implementation of a generator."""
+
+    def __init__(
+        self,
+        resources_path: str,
+        model_type: str,
+        model_name: str,
+        scheduler_type: str,
+        auth_token: bool = True,
+        prompt: Optional[str] = None,
+        device: Optional[Union[torch.device, str]] = None,
+    ):
+        """A Diffusers generation algorithm.
+
+        Args:
+            resources_path: path to the cache.
+            model_type: type of the model.
+            model_name: name of the model weights/version.
+            scheduler_type: type of the schedule.
+            auth_token: authentication token for private models.
+            prompt: target text to use for conditional generation.
+            device: device where the inference
+                is running either as a dedicated class or a string. If not provided is inferred.
+        """
+        self.device = device_claim(device)
+        self.resources_path = resources_path
+        self.model_type = model_type
+        self.model_name = model_name
+        self.scheduler_type = scheduler_type
+        self.prompt = prompt
+        self.auth_token = auth_token
+        self.load_model()
+
+    def load_model(self) -> None:
+        """Load a pretrained diffusion generative model."""
+
+        try:
+            model_class = MODEL_TYPES[self.model_type]
+        except KeyError:
+            raise KeyError(f"model type: {self.model_type} not supported")
+
+        if (
+            os.path.exists(self.resources_path)
+            and len(os.listdir(self.resources_path)) > 0
+        ):
+            model_name_or_path = self.resources_path
+        else:
+            model_name_or_path = self.model_name
+
+        if self.model_type == "stable_diffusion":
+            self.model = model_class.from_pretrained(
+                model_name_or_path,
+                use_auth_token=self.auth_token,
+            )
+        else:
+            self.model = model_class.from_pretrained(model_name_or_path)
+        self.model.to(self.device)
+
+    def sample(
+        self, batch_size: int = 1, prompt: Union[str, List[str]] = None
```

Comment:
Two comments:
- why not removing the prompt at this point? no strong feelings, since the implementation is hidden
- we should rename batch_size to number of samples, what do you think?

Reply:
Good points. Done.